doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -0,0 +1,269 @@
+   - The logic for `KeyMapping::SetKeyBinding` and `KeyMapping::ClearKeyBinding` can be moved to `Action`.
+ - Key Chord text
+   - `String Action::KeyChordText{ get; }` can be exposed to pass the text directly to the command palette.
+       This would depend on `Keys` and, thus, propagate changes automatically.
+ - Observable properties
+   - Several members can be exposed as observable (as they are in `Command`) such as the name, key chord text, and icon.
+ - Nested and iterable commands
+   - `HasNestedCommands`, `NestedCommands{ get; }`, `IterateOn` will continue to be exposed.
+   - A setter for these customizations will not be exposed until we find it necessary (i.e. adding support for customizing it in the Settings UI)
+   - Command expansion can continue to be exposed here to reduce implementation cost.
+
+Overall, the new `Action` class will be an evolution of the `Command` class that now also includes the `KeyChord` it has.
+ This allows the implementation cost of this step to be relatively small. In fact, the class itself may not even be renamed.
+
+Completion of this step should only cause relatively minor changes to anything that depends on `Command`, because
+ it is largely the same class. However, key bindings will largely be impacted because we represent key bindings as
+ a map of `KeyChord`s to `ActionAndArgs`. This leads us to step 2 of this process.
+
+
+### Step 2: Querying actions
+
+Key bindings and commands are deserialized by basically storing individual actions to a map.
+- `KeyMapping` is basically an `IMap<KeyChord, ActionAndArgs>` with a few extra functions. In fact, it actually
+ stores key binding data to a `std::map<KeyChord, ActionAndArgs>` and directly interacts with it.
+- `Command::LayerJson` populates an `IMap<String, Command>` during deserialization as it iterates over every action.
+ Note that `Command` can be interpreted as a wrapper for `ActionAndArgs` with more stuff here.
+
+It makes sense to store these actions as maps. So, following step 1 above, we can also store and expose actions
+ something like the following:
+
+```c++
+runtimeclass ActionMap
+{
+    Action GetActionByName(String name);
+    Action GetActionByKeyChord(KeyChord keys);
+
+    KeyChord GetKeyBindingForAction(ShortcutAction action);
+    KeyChord GetKeyBindingForAction(ShortcutAction action, IActionArgs actionArgs);
+
+    // Future Considerations:
+    // - [#6899]: Action IDs --> GetActionByID()
+    // - Getters for groups of actions...
+    //   - IMap<> GetActionsByGrouping
+    //   - IMap<> GetActionsByTag
+}
+```
+
+The getters will return null if a matching action or key chord is not found. Internally, we can
+ store the actions as follows:
+
+```c++
+std::map<std::wstring, Action> _NameMap;
+std::map<KeyChord, Action> _KeyMap;
+```
+
+`GetActionByName` and `GetActionByKeyChord` will directly query the internal maps.
+`GetKeyBindingForAction` will iterate through the `_KeyMap` to find a match (similar to how it is now).
+

```

Comment:
Added a layering section for the defaults.json and settings.json layering.

In an event like you described above where both occur in the same layer, my plan is to keep "copy" in the action list, but actually have "paste" be bound to ctrl+a. Serialization-wise, we'll only output:
```js
{ "command": "copy"}
{ "keys": "ctrl+a", "command": "paste"}
```

I _think_ this is what our current actions model does, because we just iterate over all of our actions and add them to the command palette independently of their key chord.